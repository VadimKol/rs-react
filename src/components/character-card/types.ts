export interface CharacterState {
  isImageLoading: boolean;
}
