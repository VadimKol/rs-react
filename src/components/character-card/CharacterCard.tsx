import Link from 'next/link';
import { useRouter } from 'next/router';
import { type ReactNode } from 'react';
import { type Character } from 'rickmortyapi';

import { FavoriteButton } from '../favorite-button/FavoriteButton';
import { ImageBlock } from '../image-block/ImageBlock';
import styles from './styles.module.scss';

export function CharacterCard({ character }: { character: Character }): ReactNode {
  const {
    query: { page, search },
  } = useRouter();

  return (
    <div className={styles.card}>
      <ImageBlock src={character.image} alt="Character" />
      <div className={styles.text_container}>
        <Link
          href={`/character/${character.id}?page=${String(page || '1')}&search=${String(search || '')}`}
          className={styles.card_link}
        >
          <h2 className={styles.card_title}>{character.name}</h2>
        </Link>
        <FavoriteButton character={character} />
      </div>
    </div>
  );
}
