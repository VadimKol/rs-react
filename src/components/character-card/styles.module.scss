.card {
  height: 100%;
  background:
    linear-gradient(var(--background-color) 0 0) padding-box,
    linear-gradient(to top, var(--primary-color), var(--secondary-color)) border-box;
  border: 5px solid transparent;
  border-radius: 10px;
}

.text_container {
  display: flex;
  flex-direction: column;
  gap: 10px;
  padding: 20px 30px;
  box-shadow: inset 0 5px 5px 0 #d3d3d3;
}

.card_title {
  margin: 0;
  font-size: 18px;
  font-weight: 800;
  text-align: center;
}

.desc {
  display: flex;
  flex-direction: column;
  gap: 10px;
  margin: 0;
  padding: 0;
  list-style-type: none;
}

.desc_item,
.type {
  margin: auto;
}

.episodes {
  display: -webkit-box;
  overflow: hidden;
  margin: 0 auto;
  text-align: left;
  text-overflow: ellipsis;
  -webkit-line-clamp: 5;
  -webkit-box-orient: vertical;
}
