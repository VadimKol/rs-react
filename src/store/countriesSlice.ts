import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

import { countries } from '@/common/countries';

import { type RootState } from './store';

export const countriesLice = createSlice({
  name: 'countries',
  initialState: countries,
  reducers: {
    addCountry: (state, action: PayloadAction<string>) => {
      state.push(action.payload);
    },
  },
});

export const { addCountry } = countriesLice.actions;

export const selectCountries = (state: RootState): string[] => state.countries;

export const useCountries = (): string[] => useSelector(selectCountries);
