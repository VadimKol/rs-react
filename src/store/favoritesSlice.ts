/* import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';
import { type Character } from 'rickmortyapi';

import { type RootState } from './store';

const initialState: Character[] = [];

export const favoritesSLice = createSlice({
  name: 'favorites',
  initialState,
  reducers: {
    addToFavorites: (state, action: PayloadAction<Character>) => {
      state.push(action.payload);
    },
    removeFromFavorites: (state, action: PayloadAction<Character>) =>
      state.filter((character) => character.id !== action.payload.id),
    clearFavorites: () => [],
  },
});

export const { addToFavorites, removeFromFavorites, clearFavorites } = favoritesSLice.actions;

const selectFavorites = (state: RootState): Character[] => state.favorites;

const isFavorite = (state: RootState, character?: Character): boolean =>
  state.favorites.some((favorite) => favorite.id === character?.id);

export const useFavorites = (): Character[] => useSelector(selectFavorites);

export const useIsFavorite = (character?: Character): boolean =>
  useSelector((state: RootState) => isFavorite(state, character));
 */
