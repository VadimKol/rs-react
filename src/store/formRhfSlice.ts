import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

import { type FormInfo } from '@/common/types';

import { type RootState } from './store';

const initialState: FormInfo = {
  name: null,
  age: null,
  email: null,
  password: null,
  confirmPassword: null,
  gender: null,
  country: null,
  tc: null,
  image: null,
  strength: '0',
};

export const formRhfSLice = createSlice({
  name: 'formRhf',
  initialState,
  reducers: {
    updateFormRhf: (_, action: PayloadAction<FormInfo>) => action.payload,
  },
});

export const { updateFormRhf } = formRhfSLice.actions;

const selectFormRhfInfo = (state: RootState): FormInfo => state.formRhf;

export const useFormRhfInfo = (): FormInfo => useSelector(selectFormRhfInfo);
