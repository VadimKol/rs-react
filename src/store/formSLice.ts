import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

import { type FormInfo } from '@/common/types';

import { type RootState } from './store';

export const initialState: FormInfo[] = [];

export const formSLice = createSlice({
  name: 'form',
  initialState,
  reducers: {
    updateForm: (state, action: PayloadAction<FormInfo>) => {
      state.push(action.payload);
    },
  },
});

export const { updateForm } = formSLice.actions;

const selectFormInfo = (state: RootState): FormInfo[] => state.form;

export const useFormInfo = (): FormInfo[] => useSelector(selectFormInfo);
