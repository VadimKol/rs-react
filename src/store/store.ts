import { configureStore } from '@reduxjs/toolkit';

import { formSLice } from './formSLice';

export const store = configureStore({ reducer: { form: formSLice.reducer } });

export type RootState = ReturnType<typeof store.getState>;
