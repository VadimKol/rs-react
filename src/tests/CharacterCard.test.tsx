import { createRemixStub } from '@remix-run/testing';
import { render } from '@testing-library/react';
import { type ReactNode } from 'react';
import { Provider } from 'react-redux';

import { CharacterCard } from '@/components/character-card/CharacterCard';
import { store } from '@/store/store';

import { character } from './mocks/data';

describe('CharacterCard Component', () => {
  it('renders correctly', () => {
    const RemixStub = createRemixStub([
      {
        path: '/',
        Component: (): ReactNode => (
          <Provider store={store}>
            <CharacterCard character={character} />
          </Provider>
        ),
      },
    ]);

    const { container } = render(<RemixStub />);

    expect(container).toMatchSnapshot();
  });
});
