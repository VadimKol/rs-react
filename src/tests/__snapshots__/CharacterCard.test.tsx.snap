// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CharacterCard Component > renders correctly 1`] = `
<div>
  <div
    class="_card_0086a2"
  >
    <div
      class="_image_container_de2c00"
    >
      <div
        class="_loader_d8e6ec"
      />
      <img
        alt="Character"
        class="_hidden_de2c00"
        height="300"
        src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
        width="300"
      />
    </div>
    <div
      class="_text_container_0086a2"
    >
      <a
        class="_card_link_0086a2"
        data-discover="true"
        href="/character/1?page=1&search="
      >
        <h2
          class="_card_title_0086a2"
        >
          Rick
        </h2>
      </a>
      <button
        aria-label="Add to favorites"
        class="_favorite_99b3bf"
        name="favorite"
        type="button"
        value="false"
      >
        â˜†
      </button>
    </div>
  </div>
</div>
`;
